3_WordBreak.cpp

```
@@ -0,0 +1,143 @@
+m = length of word dictionary
+
+3.2
+Iterate over the entire length of the string character by character. If no match is found with the words in the dictionary even when you reach the end of the string, return false.
+If you found a match, repeat this same algorithm for the remaining portion of the input string (i.e. recursive call)
+
+Complexity
+Time: O(2^n)
+Space: O(n^2)
+
+4. O - OPTIMISE
+As we can see, the Word Break problem has an optimal substructure. Also, it can be broken down into multiple subproblems.
+Hence, we can use Dynamic Programming to solve this problem wherein the subproblems will be memoised rather than being computed again and again.
+-> Use an array for storing the results of the subproblems.
+-> -1: not encountered, 0: encountered, 1: can be segmented i.e, one or more of the word breaks of the string matches words in the dictionary.
+-> Consider all prefixes of the current string, If prefix is found in dictionary, then recurse for suffix.
+5. W - WALK THROUGH
+Explained as comments in the code
+6. I - IMPLEMENT
+*/
+#include <iostream>
+#include <string>
+#include <vector>
+#include <algorithm>
+using namespace std;
+bool wordBreak(vector<string> const &dict, string str, vector<int> &canBeSegmented)
+{
+    int len = str.size();
+    if (len == 0)
+        return true;
+    //subproblem encountered for the first time
+    if (canBeSegmented[len] == -1)
+    {
+        //mark subproblem as seen
+        canBeSegmented[len] = 0;
+        for (int i = 1; i <= len; i++)
+        {
+            // consider all prefixes of current string
+            string prefix = str.substr(0, i);
+            // if prefix is found in dictionary, then recurse for suffix
+            if (find(dict.begin(), dict.end(), prefix) != dict.end() && wordBreak(dict, str.substr(i), canBeSegmented))
+                return canBeSegmented[len] = 1;
+
+        }
+    }
+    // return solution to current sub-problem
+    return canBeSegmented[len];
```

you're returning an int as a bool here, it looks like it should always be the case that the value is 0 or 1, however, it could be easy to make small changes to the code where the value would be -1 which would silently return true

---

Thanks for pointing it out! I've changed the return type of the function to int.